feat: introduce tracing

eserilev opened the pull request:
## Description

We replace `log` with `tracing` across the codebase. Where possible, we make use of structured logging now instead of templating strings.

`tracing` offers the ability to also record "span"s. A span lasts until its dropped and describes the entire duration that it is active for. All logs (in `tracing` term "events") are hierarchically embedded in all parent-spans).

We introduce several spans:

- On debug level: One for `new_outgoing_connection`, `new_incoming_connection` and `new_established_connection`
- On debug level: `Connection::poll`, `Swarm::poll` and `Pool::poll`
- On trace level: `NetworkBehaviour::poll` for each implementation of `NetworkBehaviour`
- On trace level: `ConnectionHandler::poll` for each implementation of (protocol) `ConnectionHandler`s

The idea here is that logging on debug level gives you a decent overview of what the system is doing. You get spans for the duration of connections and how often each connection gets polled.

Dropping down to trace level gives you an extremely detailed view of how long each individual `ConnectionHandler` was executed as part of `poll` which could be used for detailed analysis on how busy certain handlers are.

Most importantly, simply logging on `info` does not give you any spans. We consider `info` to be a good default that should be reasonably quiet.

Resolves #1533.

## Notes & open questions


## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

eserilev committed:
feaat:tracing

thomaseizinger commented on swarm/Cargo.toml:
```diff
@@ -31,6 +31,8 @@ multistream-select = { workspace = true }
 async-std = { version = "1.6.2", optional = true }
 tokio = { version = "1.29", features = ["rt"], optional = true }
+tracing = "0.1.37"
+tracing-subscriber = { version = "0.3", features = ["env-filter", "json"] }
```
We shouldn't need this in the library crates, right?

eserilev replied:
you're right, `tracing-subscriber` is not necessary here

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -307,6 +307,8 @@ where
 {
     /// Creates a new empty `Pool`.
     pub(crate) fn new(local_id: PeerId, config: PoolConfig) -> Self {
+        let span = tracing::error_span!("peer id:", %local_id, connection_id = tracing::field::Empty);
```
The way spans work is that they need to be within the stack of all function calls in which logged messages should contain this context. Think about it as "invisible" function parameters.

The `new` function does not have an `tracing::log` statements so this span does not nothing.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -495,6 +497,10 @@ where
     ) {
         let connection = connection.extract();
 
+        let span = tracing::error_span!("peer id:", %obtained_peer_id, connection_id = tracing::field::Empty);
```
This span also doesn't really do anything because there are no log messages following this one.

I think what we should do is construct a span in here and pass it to `Connection`. `Connection` can then call `span.enter()` on every `poll` invocation which should add the context to all `tracing` statements within every `ConnectionHandler` implementation.

thomaseizinger replied:
```suggestion
        let span = tracing::error_span!("connection", %peer_id, connection_id = tracing::field::Empty);
```

I think the name of the span should be `connection`.

Plus, it would be good to record the other peer's IP address as well I'd say.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -495,6 +497,10 @@ where
 
+        let span = tracing::error_span!("peer id:", %obtained_peer_id, connection_id = tracing::field::Empty);
+        let _guard = span.enter();
+        span.record("connection id", tracing::field::display(id));
```
You already have the connection id available when you construct the swarm, no need to use `tracing::field::Empty`.

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -172,10 +172,10 @@ fn main() -> Result<(), Box<dyn Error>> {
         .unwrap();
     swarm
-        .listen_on("/ip4/0.0.0.0/tcp/0".parse().unwrap())
+        .listen_on("/ip6/::/udp/4001/quic-v1".parse().unwrap())
```
Is this related to your change?

eserilev replied:
no, this is now removed

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -258,11 +258,11 @@ impl ConnectionHandler for Handler {
                 }
                 Poll::Ready(Ok(stream)) => {
-                    log::trace!("answered inbound ping from {}", self.peer);
+                    tracing::trace!("answered inbound ping from {}", self.peer);
```
I wonder whether we should remove `self.peer` from log statements like this.

On the one hand, for users that use `tracing`, this will be redundant information. On the other hand, non-users of `tracing` will be missing this context.

I am tempted to take a more opinionated approach here and say that, if you want informative logs, you should use `tracing` and we make those logs less noisy by removing this redundant information.

Thoughts @mxinden?

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -273,7 +273,7 @@ impl ConnectionHandler for Handler {
             // Check for outbound ping failures.
             if let Some(error) = self.pending_errors.pop_back() {
-                log::debug!("Ping failure: {:?}", error);
+                tracing::debug!("Ping failure: {:?}", error);
```
I believe tracing has a way of capturing errors, we should use that too!

eserilev replied:
updated to `tracing::error!` so that this event gets captured at the error level

thomaseizinger replied:
Sorry for not being more clear here, that is not what I meant :)

Logging this is as an `tracing::error!` is way too verbose, we don't want to wake an ops people at 2am just because a ping failed. I was referring to capturing the error as a field:

```rust
tracing::debug!(error = error as &dyn std::error::Error, "Ping failure");
```

For this, you'll also have to change `Some(error)` to `Some(ref error)`.

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -295,7 +295,7 @@ impl ConnectionHandler for Handler {
                     }
                     Poll::Ready(Ok((stream, rtt))) => {
-                        log::debug!("latency to {} is {}ms", self.peer, rtt.as_millis());
+                        tracing::debug!("latency to {} is {}ms", self.peer, rtt.as_millis());
```
Statements like this should use structured logs by capturing the RTT as a field instead of embedding it into the message: https://docs.rs/tracing/0.1.38/tracing/#recording-fields

eserilev replied:
updated to this:
` tracing::event!(tracing::Level::TRACE, peer = %self.peer, ms = rtt.as_millis());`
so that we capture peer and rtt as a field. Let me know if you wanted something different here

thomaseizinger replied:
Yep, that makes sense. You can continue to use `tracing::debug!` though, no need to for the verbose `event!` macro. We can change the message to something like: `Successfully pinged peer`.

And with the newly introduced span inside `Connection`, I don't think we need to log the `PeerId` anymore right? It should show up automatically now :)

thomaseizinger reviewed (commented):
Thank you for starting this!

I've left some notes :)

eserilev committed:
move error span

eserilev committed:
remove tracing-subscriber from library crate

thomaseizinger commented on examples/ping-example/Cargo.toml:
```diff
@@ -10,3 +10,5 @@ async-std = { version = "1.12", features = ["attributes"] }
 futures = "0.3.28"
 libp2p = { path = "../../libp2p", features = ["async-std", "dns", "macros", "noise", "ping", "tcp", "websocket", "yamux"] }
+tracing = { version = "0.1.37", features = ["log"] }
+tracing-subscriber = { version = "0.3", features = ["env-filter", "json"] }
```
Do we need the `json` feature?

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -513,11 +512,13 @@ where
         }
 
-        let connection = Connection::new(
+        let span = tracing::error_span!("connection", connection_id = %id, peer_id = %obtained_peer_id, peer_address = ?endpoint);
```
I'd prefer to not debug-print the `endpoint` but instead call `get_remote_address` on it an display-print that one.

thomaseizinger replied:
```suggestion
        let span = tracing::error_span!("connection", id = %id, peer = %obtained_peer_id, peer_address = ?endpoint);
```

Let's shorten this, otherwise it is super noisy.

thomaseizinger commented on swarm/src/connection.rs:
```diff
@@ -753,11 +767,13 @@ mod tests {
 
     #[test]
     fn propagates_changes_to_supported_inbound_protocols() {
+        let span = tracing::error_span!("test");
```
You can just call `tracing::Span::none()`.

thomaseizinger commented on swarm/src/connection.rs:
```diff
@@ -176,6 +184,7 @@ where
         mut handler: THandler,
         substream_upgrade_protocol_override: Option<upgrade::Version>,
         max_negotiating_inbound_streams: usize,
+        error_span: Span,
```
`error_span` isn't a particularly good name, perhaps just say `span`.

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -295,8 +295,7 @@ impl ConnectionHandler for Handler {
                     Poll::Ready(Ok((stream, rtt))) => {
-                        log::debug!("latency to {} is {}ms", self.peer, rtt.as_millis());
-
+                        tracing::event!(tracing::Level::TRACE, peer = %self.peer, ms = rtt.as_millis());
```
`Duration` actually has a very nice looking `Debug` impl so I'd suggest we use that:

```suggestion
                        tracing::event!(tracing::Level::TRACE, peer = %self.peer, ?rtt);
```

thomaseizinger reviewed (commented):
Thanks, good progress so far :)

The span is working, see the following screenshot from a local run of the ping-example on my machine:

![image](https://github.com/libp2p/rust-libp2p/assets/5486389/8fc73469-6bc8-454b-b6d2-b4e8c10d384b)

As you can see, some of the information is duplicated. Now that we capture the `PeerId` within the `connection` span, it is printed redundantly in some of the messages. I think we should remove it from those messages.

Also, as I noted inline the `peer_address` prints the entire `ConnectedPoint` which is very noisy.

I'd encourage you to run the example yourself (I used `RUST_LOG=libp2p_ping=trace,debug cargo run`) and tweak the logs until they are concise and informative.

Once the ping protocol is done, we can move on to other protocols and tidy up the logs there!

eserilev committed:
update logging

eserilev committed:
remove reduncdant logs

eserilev committed:
resolve conflict

mxinden commented on protocols/ping/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ libp2p-identity = { workspace = true }
 log = "0.4.19"
 rand = "0.8"
 void = "1.0"
+tracing = "0.1.37"
```
```suggestion
tracing = "0.1.37"
void = "1.0"
```

Preference of these to be in alphabetical order.

mxinden commented on swarm/Cargo.toml:
```diff
@@ -31,6 +31,7 @@ multistream-select = { workspace = true }
 [target.'cfg(not(any(target_os = "emscripten", target_os = "wasi", target_os = "unknown")))'.dependencies]
 async-std = { version = "1.6.2", optional = true }
 tokio = { version = "1.29", features = ["rt"], optional = true }
+tracing = "0.1.37"
```
I am assuming that `tracing` does have WASM support and that it should really go into the `[dependencies]` section. Is that correct?

mxinden reviewed (commented):
Excited for this to land. Thank you for working on it.

Feel free to ignore comments below at this point.

eserilev committed:
autonat tracing

eserilev committed:
dcutr changes

eserilev committed:
relay server

eserilev committed:
floodsub

eserilev committed:
gossipsub

eserilev committed:
identity

eserilev committed:
merge conflicts

eserilev committed:
kad

eserilev committed:
mdns

eserilev committed:
perf

eserilev committed:
rendezvous

eserilev committed:
request-response

eserilev committed:
chat example changes

eserilev committed:
distributed key value store example

thomaseizinger commented on swarm/src/connection.rs:
```diff
@@ -712,14 +717,15 @@ mod tests {
 
             let alive_substream_counter = Arc::new(());
-
+            let span = tracing::Span::none();
```
I think we can inline these usages of `Span::none` to condense the code a bit further :)

thomaseizinger reviewed (commented):
This looks good to me so far! I think we will have to "just" merge this eventually and improve the logs over time as we get feedback or notice things ourselves :)

eserilev committed:
inline tracing spans

eserilev committed:
add tracing to the rest of the examples

thomaseizinger commented on examples/identify/Cargo.toml:
```diff
@@ -10,3 +10,5 @@ async-std = { version = "1.12", features = ["attributes"] }
 futures = "0.3.28"
 libp2p = { path = "../../libp2p", features = ["async-std", "dns", "dcutr", "identify", "macros", "noise", "ping", "relay", "rendezvous", "tcp", "tokio", "yamux"] }
+tracing = { version = "0.1.37", features = ["log"] }
+tracing-subscriber = { version = "0.3", features = ["env-filter"] }
```
Can you please fix these missing newlines? This shows up in several files.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -40,6 +40,7 @@ use std::{
     task::{Context, Poll},
     time::Duration,
 };
+use tracing::{debug, trace, warn};
```
Please no imports of these macros but just fully-qualified usages :)

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -295,8 +295,7 @@ impl ConnectionHandler for Handler {
                     Poll::Ready(Ok((stream, rtt))) => {
-                        log::debug!("latency to {} is {}ms", self.peer, rtt.as_millis());
-
+                        tracing::event!(tracing::Level::TRACE, ?rtt);
```
```suggestion
                        tracing::debug!(?rtt, "ping succeeded");
```

I think adding _some_ message here makes sense.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -513,11 +512,13 @@ where
         }
 
-        let connection = Connection::new(
+        let span = tracing::error_span!("connection", id = %id, peer = %obtained_peer_id, peer_address = %endpoint.get_remote_address());
```
```suggestion
        let span = tracing::error_span!("connection", id = %id, peer = %obtained_peer_id, remote_address = %endpoint.get_remote_address());
```

thomaseizinger replied:
Thinking about it more, we should call this span after the method that is being called: `Connection::poll`.

eserilev replied:
correct me if I'm wrong, but i think line 247 in `connection.rs` already does what you described

thomaseizinger commented on protocols/gossipsub/src/mcache.rs:
```diff
@@ -21,7 +21,7 @@
 use crate::types::{MessageId, RawMessage};
 use libp2p_identity::PeerId;
-use log::{debug, trace};
+use tracing::{debug, trace};
```
Let's tidy these up whilst we touch them and not import these macros anywhere but fully qualify them. Thanks! :)

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -183,7 +183,7 @@ impl Behaviour {
         for p in peers {
             if !self.connected.contains_key(&p) {
-                log::debug!("Not pushing to {p} because we are not connected");
+                tracing::debug!("Not pushing to {p} because we are not connected");
```
Not a must but if you are feeling like it, statements like these could be re-written to take advantage of structured logging, e.g.:

```suggestion
                tracing::debug!(peer = %p, "Cannot push to a peer we are not connected to");
```

thomaseizinger replied:
There are likely dozens of such opportunities in the codebase :)

Let me know how far you want to take this!

eserilev replied:
just to make sure I'm understanding this correctly

would we want something like this
`tracing::info!("Discovered peer {} at {}", peer, address);`
to be converted to
`tracing::info!(peer=%peer,  "Discovered peer at {}", address);`

thomaseizinger replied:
Or even better:

```rust
tracing::info!(%peer, %address, "Discovered peer");
```

thomaseizinger reviewed (commented):
Thanks! A few more comments :)

eserilev committed:
merge conflicts

eserilev committed:
fully qualified tracing usage

eserilev committed:
fix

eserilev committed:
fix cargo newline issue

eserilev committed:
fully qualify trace usage

thomaseizinger commented on interop-tests/src/bin/wasm_ping.rs:
```diff
@@ -14,7 +14,7 @@ use tokio::process::Child;
 use tower_http::cors::CorsLayer;
 use tower_http::trace::TraceLayer;
-use tracing::{error, warn};
+use tracing;
```
Shouldn't need this because that is the crate name! If you run clippy, it should actually tell you :)

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Fix clippy lints

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Fix wasm compile error

eserilev committed:
added improved tracing for autonat, dcutr, floodsub, gossipsub

eserilev committed:
tracing improvements to kad and identify

eserilev committed:
mdns tracing improvements

eserilev committed:
perf tracing improvements

eserilev committed:
the rest of the protocols

eserilev committed:
remove tracing from core crate

eserilev committed:
removed log dependency from identity crate

eserilev committed:
swap log w/ trace for interop-tests

eserilev committed:
browser-webrtc example tracing

eserilev committed:
dcutr example tracing

eserilev committed:
memory-connection-limits tracing

eserilev committed:
multistream-select tracing

eserilev committed:
server tracing

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -526,14 +526,14 @@ where
     /// subscribed.
     pub fn subscribe<H: Hasher>(&mut self, topic: &Topic<H>) -> Result<bool, SubscriptionError> {
-        tracing::debug!("Subscribing to topic: {}", topic);
+        tracing::debug!(%topic, "Subscribing to topic:");
```
Let's remove the `:`.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -526,14 +526,14 @@ where
 
         if self.mesh.get(&topic_hash).is_some() {
-            tracing::debug!("Topic: {} is already in the mesh.", topic);
+            tracing::debug!(%topic, "Topic is already in the mesh.");
```
Let's remove the `.` here.

```suggestion
            tracing::debug!(%topic, "Topic is already in the mesh");
```

thomaseizinger commented on protocols/gossipsub/src/peer_score.rs:
```diff
@@ -274,11 +274,11 @@ impl PeerScore {
+                        %deficit,
+                        "[Penalty] The peer has a mesh message deliveries deficit in topic\
+                        and will get penalized by {}",
                         p3 * topic_params.mesh_message_deliveries_weight
```
Let's make this value a structured log too! The calculation is in fact duplicated so we can extract it in a variable, see line 272.

thomaseizinger replied:
Let's call it the computed variable `penalty`.

thomaseizinger commented on protocols/gossipsub/src/peer_score.rs:
```diff
@@ -274,11 +274,11 @@ impl PeerScore {
+                        %topic,
+                        %deficit,
+                        "[Penalty] The peer has a mesh message deliveries deficit in topic\
+                        and will get penalized by {}",
```
```suggestion
                        "[Penalty] The peer has a mesh message deliveries deficit and will penalized"
```

thomaseizinger commented on protocols/gossipsub/src/peer_score.rs:
```diff
@@ -347,8 +348,8 @@ impl PeerScore {
-                "[Penalty] Behavioral penalty for peer {}, count = {}.",
-                peer_id,
+                peer=%peer_id,
+                "[Penalty] Behavioral penalty for peer, count = {}.",
```
Let's make `count` a value in the log too.

thomaseizinger commented on protocols/gossipsub/src/peer_score.rs:
```diff
@@ -594,7 +595,12 @@ impl PeerScore {
+                peer=%from,
+                status=?record.status,
+                "Unexpected delivery trace: Message from peer was first seen {}s ago", 
+                record.first_seen.elapsed().as_secs(),
```
`Duration`s are actually best logged via `Debug`! That implements a human-friendly formatter that computes the correct precision based on the value.

Let's make that a field too please!

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -372,8 +372,8 @@ impl ConnectionHandler for Handler {
-                        "Supported listen protocols changed from [{before}] to [{after}], pushing to {}",
-                        self.remote_peer_id
+                        peer=%self.remote_peer_id,
+                        "Supported listen protocols changed from [{before}] to [{after}], pushing to peer"
```
Let's extract these as values too!

thomaseizinger commented on protocols/kad/src/handler.rs:
```diff
@@ -704,11 +708,17 @@ impl ConnectionHandler for KademliaHandler {
-                            "Now operating in {new_mode}-mode on outbound connection with {peer}"
+                            %peer,
+                            mode=%new_mode,
+                            "Now operating in mode on outbound connection with peer"
```
```suggestion
                            "Changed mode on outbound connection"
```

thomaseizinger commented on protocols/kad/src/handler.rs:
```diff
@@ -704,11 +708,17 @@ impl ConnectionHandler for KademliaHandler {
+                            %peer,
+                            mode=%new_mode,
+                            local_address=%local_addr,
+                            "Now operating in mode on inbound connection with peer assuming that one of our external addresses routes to the local address")
```
Similar re-wording as above please.

thomaseizinger commented on protocols/kad/src/handler.rs:
```diff
@@ -481,12 +481,16 @@ impl KademliaHandler {
-                    "Operating in {mode}-mode on new outbound connection to {remote_peer_id}"
+                    peer=%remote_peer_id,
+                    mode=%mode,
+                    "Operating in mode on new outbound connection to peer"
```
```suggestion
                    "New outbound connection"
```

thomaseizinger commented on protocols/kad/src/handler.rs:
```diff
@@ -481,12 +481,16 @@ impl KademliaHandler {
-                    "Operating in {mode}-mode on new inbound connection to {remote_peer_id}"
+                    peer=%remote_peer_id,
+                    mode=%mode,
+                    "Operating in mode on new inbound connection to peer"
```
```suggestion
                    "New inbound connection"
```

thomaseizinger commented on protocols/perf/src/bin/perf.rs:
```diff
@@ -131,7 +131,7 @@ async fn server(server_address: SocketAddr) -> Result<()> {
                     peer_id, endpoint, ..
                 } => {
-                    tracing::info!("Established connection to {:?} via {:?}", peer_id, endpoint);
+                    tracing::info!(peer=%peer_id, ?endpoint, "Established connection to peer via endpoint");
```
```suggestion
                    "Established new connection"
```

thomaseizinger commented on core/Cargo.toml:
```diff
@@ -30,6 +29,7 @@ rw-stream-sink = { workspace = true }
 serde = { version = "1", optional = true, features = ["derive"] }
 smallvec = "1.11.0"
 thiserror = "1.0"
+tracing = { version = "0.1.37", features = ["log"] }
```
I think activating `log` is something that downstream users should do!

eserilev replied:
got it, ill update relevant tracing imports to remove the `log` feature

thomaseizinger commented on core/src/transport/global_only.rs:
```diff
@@ -292,20 +291,20 @@ impl<T: crate::Transport + Unpin> crate::Transport for Transport<T> {
             Some(Protocol::Ip4(a)) => {
                 if !ipv4_global::is_global(a) {
-                    debug!("Not dialing non global IP address {:?}.", a);
+                    tracing::debug!(ip=?a, "Not dialing non global IP address.");
```
Can we display-fmt the address?

thomaseizinger commented on core/src/transport/global_only.rs:
```diff
@@ -292,20 +291,20 @@ impl<T: crate::Transport + Unpin> crate::Transport for Transport<T> {
             Some(Protocol::Ip6(a)) => {
                 if !ipv6_global::is_global(a) {
-                    debug!("Not dialing non global IP address {:?}.", a);
+                    tracing::debug!(ip=?a, "Not dialing non global IP address.");
```
Here too.

thomaseizinger commented on core/src/transport/global_only.rs:
```diff
@@ -292,20 +291,20 @@ impl<T: crate::Transport + Unpin> crate::Transport for Transport<T> {
             }
             _ => {
-                debug!("Not dialing unsupported Multiaddress {:?}.", addr);
+                tracing::debug!(address=%addr, "Not dialing unsupported Multiaddress.");
```
Let's remove dots at the end while we are at it for the log statements that we touch!

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -258,31 +257,31 @@ fn main() -> Result<(), Box<dyn Error>> {
                     peer_id, endpoint, ..
                 } => {
-                    info!("Established connection to {:?} via {:?}", peer_id, endpoint);
+                    tracing::info!(peer=%peer_id, endpoint=?endpoint, "Established connection to peer via endpoint");
```
```suggestion
                    tracing::info!(peer=%peer_id, endpoint=?endpoint, "Established new connection");
```

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -258,31 +257,31 @@ fn main() -> Result<(), Box<dyn Error>> {
                 }
                 SwarmEvent::OutgoingConnectionError { peer_id, error, .. } => {
-                    info!("Outgoing connection error to {:?}: {:?}", peer_id, error);
+                    tracing::info!(peer=?peer_id, "Outgoing connection error to peer: {:?}", error);
```
```suggestion
                    tracing::info!(peer=?peer_id, "Outgoing connection failed: {error}");
```

thomaseizinger reviewed (commented):
Awesome progress!

I've left a few ideas for improvements :)

eserilev committed:
muxer crate tracing

eserilev committed:
yamux, test harness tracing

eserilev committed:
autonat tracing, plus remove unwrap

eserilev committed:
keypair tracing fix

eserilev committed:
upnp and a few other crates tracing

eserilev committed:
swarm tracing

eserilev committed:
swarm test tracing

eserilev committed:
cargo fmt

eserilev committed:
tracing cleanup

eserilev committed:
remove '.' from tracing messages

eserilev committed:
tracing dns crate, plus alphabatize dependancies

eserilev committed:
noise crate tracing

eserilev committed:
plaintext crate tracing

eserilev committed:
pnet crate tracing

eserilev committed:
added quic tracing

eserilev committed:
tcp crate tracing

eserilev committed:
webrtc crate tracing

eserilev committed:
weebrtc-websys crate tracing

eserilev committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into feat-tracing

eserilev committed:
websocket crate tracing

eserilev committed:
fix dependencies

eserilev committed:
changes based on feedback

eserilev committed:
cargo fmt

eserilev committed:
resovle conflicts

eserilev commented on misc/server/src/main.rs:
```diff
@@ -108,25 +107,25 @@ async fn main() -> Result<(), Box<dyn Error>> {
         swarm.add_external_address(address.clone())
     }
-    info!(
+    tracing::info!(
```
i decided not to use structured logging for vectors/lists

eserilev commented on protocols/autonat/src/behaviour/as_server.rs:
```diff
@@ -111,9 +111,9 @@ impl<'a> HandleInnerEvent for AsServer<'a> {
-                            peer,
+                        tracing::debug!(
+                            %peer,
+                            "Inbound dial request from peer with dial-back addresses {:?}",
```
since this is a list of addresses, I didnt use structured logging

eserilev committed:
personal review

thomaseizinger commented on examples/autonat/src/bin/autonat_server.rs:
```diff
@@ -37,7 +39,11 @@ struct Opt {
 async fn main() -> Result<(), Box<dyn Error>> {
-    env_logger::init();
+    let env_filter = EnvFilter::builder()
+        .with_default_directive(LevelFilter::DEBUG.into())
```
I think the default is `INFO`, right? Is logging on debug worth the extra boilerplate here or should we just use `EnvFilter::from_default_env()` and save a few lines in all examples?

thomaseizinger replied:
We could replace this with the following which is quite a bit shorter:

```rust
tracing_subscriber::fmt().with_env_filter(EnvFilter::from_default_env()).init();
```

thomaseizinger commented on misc/server/src/main.rs:
This binary is still using `env_logger` which won't pick up the new logs now.

thomaseizinger commented on protocols/identify/Cargo.toml:
I've not checked all crates but this one has `env_logger` still in its dev-dependencies. I think we should replace that with `tracing-subscriber` otherwise we won't see anything in the logs when we run the test and wonder why :)

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -148,7 +148,7 @@ pub struct Handler {
     state: State,
     /// The peer we are connected to.
-    peer: PeerId,
+    _peer: PeerId,
```
We can remove this field if it is no longer needed. The `Handler` is crate-private so we should be able to do that without breaking functionality.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -517,12 +516,14 @@ where
         }
 
-        let connection = Connection::new(
+        let span = tracing::error_span!("connection", id = %id, peer = %obtained_peer_id, remote_address = %endpoint.get_remote_address());
```
```suggestion
        let span = tracing::error_span!("Connection::poll", id = %id, peer = %obtained_peer_id, remote_address = %endpoint.get_remote_address());
```

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -517,12 +516,14 @@ where
 
-        let connection = Connection::new(
+        let span = tracing::error_span!("connection", id = %id, peer = %obtained_peer_id, remote_address = %endpoint.get_remote_address());
+        let connection: Connection<THandler> = Connection::new(
```
Shouldn't need the type-hint?

```suggestion
        let connection = Connection::new(
```

thomaseizinger reviewed (commented):
Thank you for making the super thorough changes! I am really keen to get this landed! Whilst technically not breaking, I do still think it is a big enough change to warrant being included in a breaking release instead of "sneaking" it in in a patch-release.

The good news is, we are very soon going to open a new merge window for breaking changes which this can go into.

I've left a few final comments, most importantly, I do think we need to also convert the tests to initialize the correct kind of logger for this to not be a degradation in experience.

There are likely several follow-up we can do here, most notably, we should experiment with creating more spans to allow users to track the interval behaviour of a `Swarm`. I have a few ideas around this that we can discuss once we've landed this!

Thanks again for this work and being so thorough despite the large blast radius! :)

eserilev committed:
merge

mxinden commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -527,14 +526,14 @@ where
     /// subscribed.
     pub fn subscribe<H: Hasher>(&mut self, topic: &Topic<H>) -> Result<bool, SubscriptionError> {
-        debug!("Subscribing to topic: {}", topic);
+        tracing::debug!(%topic, "Subscribing to topic");
```
//CC @jxs and @AgeManning to make sure you are aware of the switch to `tracing` in libp2p in general and Gossipsub in particular.

AgeManning replied:
Thanks for the ping! 

I'm in favour of tracing! I use it in other libraries :)

mxinden reviewed (commented):
Thank you for all the work.

This looks good to me. No need to block on my review.

The only slight worry I have, is that this pull request has a performance impact. Is this something to expect? If so, we can validate this on the rust-libp2p IPFS bootstrap staging server once it is in `master`.

eserilev committed:
remove env logger dependency

eserilev committed:
fmt

eserilev committed:
fixes

eserilev committed:
fmt

eserilev committed:
fmt

eserilev committed:
fixes based on feedback

eserilev committed:
remove unnecessary peer id ref

thomaseizinger commented on examples/browser-webrtc/src/main.rs:
```diff
@@ -20,12 +20,12 @@ use rand::thread_rng;
 #[tokio::main]
 async fn main() -> anyhow::Result<()> {
-    env_logger::builder()
-        .parse_filters("browser_webrtc_example=debug,libp2p_webrtc=info,libp2p_ping=debug")
```
Lets retain these please.

thomaseizinger reviewed (commented):
Great! Thank you :)

Now, the last thing that would be great is the example we talked about that shows, how to make use of the span data using open-telemetry! Do you see yourself doing that or would you like me to add that?

eserilev committed:
fixing tracing subscriber issue

eserilev committed:
fmt

eserilev committed:
add mdns + ping example

eserilev committed:
otel example

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Fix compile errors

thomaseizinger committed:
Revert bad updates

thomaseizinger committed:
Add more spans and extend metrics example

thomaseizinger committed:
Add more docs

thomaseizinger committed:
Remove span from select

thomaseizinger commented on libp2p/src/tutorials/ping.rs:
```diff
@@ -56,9 +56,9 @@
 //!        libp2p = { version = "0.52", features = ["tcp", "dns", "async-std", "noise", "yamux", "websocket", "ping", "macros"] }
-//!        futures = "0.3"
-//!        env_logger = "0.10.0"
-//!        async-std = { version = "1.12", features = ["attributes"] }
```
Whoops, bad merge.

thomaseizinger committed:
Move muxer spans into `Connection`

thomaseizinger committed:
Add spans for `Swarm`

thomaseizinger committed:
Add local peer id to `Swarm::poll` span

thomaseizinger committed:
Move spans to concrete `poll` functions

thomaseizinger committed:
Move stream muxer spans to concrete impl

thomaseizinger committed:
Remove unused imports

thomaseizinger committed:
Instrument dialing of new connections

thomaseizinger committed:
Remove parents

These spans are going to spawned into an executor so they shouldn't
prolong the span they are started in.

thomaseizinger committed:
Set `follows_from` instead

thomaseizinger committed:
Fix doc link

thomaseizinger committed:
Fix libp2p doc tests

thomaseizinger committed:
Fix bad char

thomaseizinger committed:
Merge branch 'master' into feat-tracing

eserilev committed:
merge

eserilev committed:
resolve merge conflicts

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Fix compile errors and delete otel-example

thomaseizinger committed:
Add spans for durations of connections

thomaseizinger committed:
Remove fields already present in parent span

thomaseizinger committed:
Fix link

thomaseizinger committed:
Fix doc tests and warnings

thomaseizinger committed:
Remove `local_peer_id` from `Swarm::poll`

This is too noisy.

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -225,6 +221,7 @@ impl ConnectionHandler for Handler {
 
     fn on_behaviour_event(&mut self, _: Void) {}
 
+    #[tracing::instrument(level = "trace", name = "ConnectionHandler::poll", skip(self, cx))]
```
Before we merge this, we need to apply this to all handlers of all protocols. Same for `NetworkBehaviour`.

thomaseizinger commented on protocols/ping/src/handler.rs:
```diff
@@ -254,11 +251,11 @@ impl ConnectionHandler for Handler {
                 Poll::Pending => {}
                 Poll::Ready(Err(e)) => {
-                    log::debug!("Inbound ping error: {:?}", e);
+                    tracing::error!("Inbound ping error: {:?}", e);
```
@eserilev Oh, we shouldn't upgrade these to errors. Can you please retain the log-level here (and further down too?)

mxinden commented on examples/metrics/README.md:
````diff
@@ -34,6 +34,45 @@ To run the example, follow these steps:
+export RUST_LOG=info,[ConnectionHandler::poll]=trace,[NetworkBehaviour::poll]=trace
+```
+
+Next, (re)-start the two example for it to connect to the OTEL collector.
````
Note that on the second `cargo run` I get the following error. Not sure whether that is worth fixing.

```
error binding to 127.0.0.1:8080: error creating server listener: Address already in use (os error 98)
```

thomaseizinger replied:
I think that is actually unrelated to this PR because 8080 is used for the HTTP server that serves the prometheus metrics.

mxinden commented on examples/metrics/README.md:
:pray:

mxinden commented on Cargo.lock:
```diff
@@ -4185,6 +4307,29 @@ dependencies = [
  "syn 2.0.38",
 ]
 
+[[package]]
+name = "prost"
+version = "0.11.9"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "0b82eaa1d779e9a4bc1c3217db8ffbeabaae1dca241bf70183242128d48681cd"
+dependencies = [
+ "bytes",
+ "prost-derive",
+]
+
+[[package]]
+name = "prost-derive"
+version = "0.11.9"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "e5d2d8d10f3c6ded6da8b05b5fb3b8a5082514344d56c9f871412d29b4e075b4"
+dependencies = [
+ "anyhow",
+ "itertools",
+ "proc-macro2",
+ "quote",
+ "syn 1.0.109",
+]
+
```
Bummer :/

thomaseizinger replied:
Oh really? I think that is only for the examples though, right? I think that is coming from opentelemetry.

thomaseizinger replied:
Confirmed, this is pulled in by `opentelemetry` and thus does not impact our users (unless they use `opentelemetry`).

mxinden commented on examples/chat/src/main.rs:
```diff
@@ -37,6 +38,14 @@ struct MyBehaviour {
+    let _ = tracing_subscriber::fmt()
+        .with_env_filter(env_filter)
+        .try_init();
+
```
Why are some examples initiated like the above, and others like:

``` rust
    let _ = tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .try_init();
```

thomaseizinger replied:
Probably just some copy-paste errors. Honestly, it is a bit annoying that this is so noisy now. Given that it is just in tests, should be maybe introduce a `publish = false` crate that offers some utilities around initializing tracing as part of a test?

mxinden replied:
I am fine either way. Agreed that it is noisy.

mxinden commented on swarm/src/connection/pool.rs:
```diff
@@ -425,20 +426,22 @@ where
-            dials,
-            dial_concurrency_factor_override.unwrap_or(self.dial_concurrency_factor),
-        );
+        let concurrency_factor =
+            dial_concurrency_factor_override.unwrap_or(self.dial_concurrency_factor);
+        let span = tracing::debug_span!(parent: tracing::Span::none(), "new_outgoing_connection", %concurrency_factor, num_dials=%dials.len(), id = %connection_id);
+        span.follows_from(tracing::Span::current());
```
:rocket:

mxinden reviewed (commented):
Apart from the outstanding comments, this is good to merge from my end. Thank you for the work! Looking forward to the insights we will gain from this.

eserilev committed:
dont upgrade debug to error

eserilev committed:
add tracing decorator to handlers

eserilev committed:
add tracing decorator to network behaviours

eserilev committed:
merge

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Remove unused imports

thomaseizinger committed:
Fix logger init in test

thomaseizinger committed:
Merge remote-tracking branch 'origin/feat-tracing' into feat-tracing

# Conflicts:
#	libp2p/src/tutorials/ping.rs

thomaseizinger committed:
Unify init of subscriber across examples

thomaseizinger committed:
Serve metrics on random port to avoid panic

thomaseizinger committed:
Fix bad merge

thomaseizinger committed:
Merge branch 'master' into feat-tracing

thomaseizinger committed:
Merge branch 'master' into feat-tracing

mxinden reviewed (approved):
Great work everyone!
